packages/@lwc/style-compiler/src/serialize.ts

```diff
@@ -165,21 +186,44 @@ function serializeCss(result: Result, collectVarFunctions: boolean): string {
             currentRuleTokens.push({ type: TokenType.text, value: part });
 
             // If we are in fakeShadow we dont want to have :host selectors
-            if ((node as any)._isHostNative) {
+            if ((node as any)._isNativeHost) {
                 // create an expression for all the tokens (concatenation of strings)
                 // Save it so in the next rule we can apply a ternary
-                tmpHostExpression = generateExpressionFromTokens(currentRuleTokens);
-            } else if ((node as any)._isFakeNative) {
-                const exprToken = generateExpressionFromTokens(currentRuleTokens);
+                nativeHostTokens = [...currentRuleTokens];
+            } else if ((node as any)._isSyntheticHost) {
+                if (!nativeHostTokens) {
+                    throw new Error('Unexpected host rules ordering');
+                }
+
+                // For `:host()` selectors, the token lists for native vs synthetic will be identical at the end of
+                // each list. So as an optimization, we can de-dup these tokens.
+                // See: https://github.com/salesforce/lwc/issues/3224#issuecomment-1353520052
+
+                const numDuplicatedTokens = calculateNumDuplicatedTokens(
+                    nativeHostTokens,
+                    currentRuleTokens
+                );
+                const numUniqueNativeTokens = nativeHostTokens.length - numDuplicatedTokens;
+                const numUniqueSyntheticTokens = currentRuleTokens.length - numDuplicatedTokens;
 
+                const uniqueNativeTokens = nativeHostTokens.slice(0, numUniqueNativeTokens);
+                const uniqueSyntheticTokens = currentRuleTokens.slice(0, numUniqueSyntheticTokens);
+
+                const nativeExpression = generateExpressionFromTokens(uniqueNativeTokens);
+                const syntheticExpression = generateExpressionFromTokens(uniqueSyntheticTokens);
+
+                // Generate a ternary to switch between native vs synthetic for the unique tokens
                 tokens.push({
                     type: TokenType.expression,
-                    value: `(${USE_ACTUAL_HOST_SELECTOR} ? ${tmpHostExpression} : ${exprToken})`,
+                    value: `(${USE_ACTUAL_HOST_SELECTOR} ? ${nativeExpression} : ${syntheticExpression})`,
                 });
 
-                tmpHostExpression = null;
+                // The remaining tokens are duplicated between native and synthetic
+                tokens.push(...currentRuleTokens.slice(numUniqueSyntheticTokens));
+
+                nativeHostTokens = undefined;
```

Comment:
Solving this in the serializer rather than the transformer is a bit ugly, but I found it easier than changing how we do the `:host()` transformation.

Reply:
How would you feel if we moved the logic for de-duplicating the token to [`post-lwc-plugin`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/style-compiler/src/postcss-lwc-plugin.ts#L54-L68)?

Since we're already comparing the results of the native vs synthetic `:host()` transformation, it feels like it would fit in here.

I was thinking we could store the duplicate portions on the nodes themselves and in `serialize.ts` we would just read the values off the nodes to generate the final host tokens.

Reply:
@jmsjtu I imagine doing it in the `postcss-lwc-plugin` would be a lot more elegant, yeah. I just did it here because I found it easier. 🙂

Could we address that in a future PR though? Now that we have the tests, we should be able to refactor later and the tests will still pass.

Reply:
Sounds good to me!